extensions/attributegen/config.proto

```diff
@@ -0,0 +1,113 @@
+//
+// If multiple AttributeGenerations produce the same attribute, the result of the last generation
+// will be visible to downstream filters.
+message AttributeGeneration {
+  // Phase denotes plugin lifecycle phase when the new attribute is generated.
+  // Default: After the response is sent back to the client.
+  Phase phase = 1;
+
+  // The following attribute is populated on a successful match.
+  // Example: `istio.operationId`
+  // `istio.` attribute namespace is reserved by Istio.
+  //
+  // AttributeGeneration may fail to evaluate when an attribute is not available.
+  // Example: `response.code` may not be available when a request ends abruptly.
+  // When evaluation fails the plugin will set an error_attribute to the error string.
+  // error_attribute = ${output_attribute}_error
+  //
+  // If the generated attribute is used by an authz plugin, it should fail the request if
+  // the `error_attribute` is set.
+  // If the generated attribute is used by Stats, it may log the error but not
+  // fail the request.
+  string output_attribute = 2;
+ 
+  // Matches are evaluated in order until the first successful match.
+  // The value specified by the successful match is assgined to the output_attribute.
+  repeated Match match = 3;
+}
+
+// If the condition evaluates to true then the Match returns the specified value. 
+message Match {
+  // Condition is a [CEL expressions](https://github.com/google/cel-spec/blob/master/doc/langdef.md)
+  // You may use [well known attributes]
+  // (https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/rbac_filter#condition)
+  // in the CEL expression.
+  //
+  // Examples:
+  //
+  // a GET on /books
+  // { value: ListBooks,
+  //   condition: request.url_path == "/books" && request.method == "GET" }
+  //
+  // a GET on /shelves/{id1}/books/{id2}
+  // { value: GetBook,
+  //   condition: request.url_path.matches("^/shelves/\d*/books/\d*$") && request.method == "GET" }
+  // Note: CEL uses [re2](https://github.com/google/re2/wiki/Syntax) regex library.
+  //       Use anchors {^, $} to ensure that the regex evaluates efficiently.
+  // Note: request.url_path is normalized and stripped of query params.
+  //
+  // a Read only operation on books
+  // { value: ReadOnlyBooks, 
+  //   condition: request.url_path.startsWith("/books/") && in(request.method, ["GET", "HEAD"])}
+  //
+  // An empty condition evaluates to `true` and should be used to provide a default value.
+  string condition = 1;
+
+  // If condition evaluates to true, return the `value`.
+  string value = 2;
+}
+
+// Phase determines when the classification is performed.
+enum Phase {
+  option allow_alias = true;
+  // AttributeGeneration is performed after the response is sent back to the client.
+  // All information about the request is available in this phase,
+  // and classification does not add to the latency observed by the client.
+  PHASE_UNSPECIFIED = 0;
+  AFTER_RESPONSE = 0;
```

Comment:
Using response/request is fine for now and will work for HTTP, how would you name it so that’s it’s suitable for other protocols?

I have used “transaction” as an abstract way to define this in the past but Istio doesn’t really use transaction in the current vocabulary.

Reply:
REQUEST, REPLY and LOG can be mapped to most other protocols.

1. memcache / redis all 3 phases are clear
2. mysql  same.
3.kafka  REQUEST (Message). and LOG are ok. There is no reply for payload.
4. grpc stream is like kafka, and bidi-grpc is like 2 kafkas... 

can you suggest something else? @kyessenov  ?

Reply:
Can we remove this option for now since we only support one callback insertion point.
I just think it's too much for a user to configure both filter & callback placements. 

Long term I was thinking for filters & extensions we should defined "classes/categories" and based on the class your placement in filter chain is more or less. This same category can also dictate the phase.

Reply:
The code supports both insertion points, it is just hidden from the docs.
I am ok hiding Phase completely if you want, but if you need to classify and use the attribute for policy decisions the *must* specify the callback not just the insertion order.

We can recast phases as filter `functions`, as it is done in the extensions API to make it less implementation specific.
This does determine the plugin ordering, but the plugin also needs to know where it will run.

```proto
enum Function {
    UNSPECIFIED_FUNCTION = 0;
    TELEMETRY = 1;
    PRE_PROCESSING = 2;
    AUTHORIZATION = 3;
    POST_PROCESSING = 4;
}
```

Reply:
Yes! These names make much more sense to an end user. With some documentation around these terms like Pre processing and post processing and examples I think it’s very reasonable and good UX.

I don’t mean to be a bottleneck on naming but I believe getting it right now will save us a lot of headaches in the future.

@dougreid thoughts?

Reply:
Unfortunately, this categorization is rather simplistic. Even the native filter mix and match functions in the same filters (filter does auth, telemetry, and pre-processing at the same time). It's a lot of work to try to re-organize existing native extensions to follow this paradigm, so I'm not sure we want to commit to that in this filter. I think we want this low-level be as close to the metal as to what actually happens in Envoy, and add some sugar on top if that irks some people the wrong way.

Reply:
Put it the other way - I don't dismiss UX, but it should reflect the reality :)

Reply:
Being close to the implementation was my goal initially since this is a filter and xds is not supposed to human consumable or producible. 

However we are caught in the middle because the xds APi *is* exposed as is to the end user for now and so UX discussion comes in.
Having a higher level abstraction like TELEMETRY, et al without the higher level extension mechanism to put things in the correct place only gets us half way.

We have 2 options
1. Stick with what we have since it maps cleanly to envoy. We can de emphasize the phase is the doc and examples.
2. Remove the option completely. but we will be in the same spot as today when someone asks for the feature in the future.

Reply:
My suggestion here would be to hide it until someone asks for it, or we have more understanding of requirements.

@kyessenov I agree about pre-mature APIs, hence why expose something that we are not sure of for now.

Reply:
Ok done, I hid the option from the proto.